### PR TITLE
Fix 32bit issues in dask.array reductions

### PR DESCRIPTION
**Pull request by mrocklin**
Fixes https://github.com/dask/dask/issues/2022

### CODE REVIEW

File: dask/array/reductions.py [hunk 1]
```diff
@@ -265,7 +265,7 @@ def nanmean(a, axis=None, dtype=None, keepdims=False, split_every=None):
 
 def moment_chunk(A, order=2, sum=chunk.sum, numel=numel, dtype='f8', **kwargs):
     total = sum(A, dtype=dtype, **kwargs)
-    n = numel(A, **kwargs)
+    n = numel(A, **kwargs).astype(np.int64)
```
Comment (mrocklin) [inline dask/array/reductions.py]: @jcrist this change confuses me.  Any idea why it might be necessary?

**Comment by mrocklin** [comment]
> Possibly accuracy/overflow issues for things where this might be int32 instead? Not sure.

The differences for computations as simple as `da.arange(10, chunks=(5,)).std()` were off by a factor of two.  It was a little bit crazy.

**Comment by jcrist** [comment]
Hmmm, I'd have to intermediates to see what's going on.

**Comment by mrocklin** [comment]
This is the example that lead me to the current solution:

### 32 bit

```python
In [1]: import dask.array as da

In [2]: x = da.arange(10, chunks=(5,))

In [3]: x.std().compute()
Out[3]: 4.4158804331639239

In [4]: x.compute().std()
Out[4]: 2.8722813232690143

In [5]: y = x.std()

In [6]: keys = list(y.dask)

In [7]: dict(zip(keys, y._get(y.dask, keys)))
Out[7]: 
{('arange-148142c4b92299810fc0c0f4549563ac', 0): array([0, 1, 2, 3, 4]),
 ('arange-148142c4b92299810fc0c0f4549563ac', 1): array([5, 6, 7, 8, 9]),
 ('moment_chunk-e036af1434ef148d014fa1bb577270f5',
  0): array([( 10., 5, [ 10.])], 
       dtype=[('total', '<f8'), ('n', '<i4'), ('M', '<f8', (1,))]),
 ('moment_chunk-e036af1434ef148d014fa1bb577270f5',
  1): array([( 35., 5, [ 10.])], 
       dtype=[('total', '<f8'), ('n', '<i4'), ('M', '<f8', (1,))]),
 ('sqrt-74b47c30b2c5723caa39612f742b39b1',): 4.4158804331639239,
 ('var-aggregate-41b80008d10fbc73f6b6222fcebbb767',): 19.5}
```

### 64 bit

```python
In [1]: import dask.array as da

In [2]: x = da.arange(10, chunks=(5,))

In [3]: x.std().compute()
Out[3]: 2.8722813232690143

In [4]: y = x.std()

In [5]: keys = list(y.dask)

In [6]: dict(zip(keys, y._get(y.dask, keys)))
Out[6]: 
{('arange-148142c4b92299810fc0c0f4549563ac', 0): array([0, 1, 2, 3, 4]),
 ('arange-148142c4b92299810fc0c0f4549563ac', 1): array([5, 6, 7, 8, 9]),
 ('moment_chunk-65a554878cd0aea424e28098ec994f45',
  0): array([(10.0, 5, [10.0])], 
       dtype=[('total', '<f8'), ('n', '<i8'), ('M', '<f8', (1,))]),
 ('moment_chunk-65a554878cd0aea424e28098ec994f45',
  1): array([(35.0, 5, [10.0])], 
       dtype=[('total', '<f8'), ('n', '<i8'), ('M', '<f8', (1,))]),
 ('sqrt-cb42bc48037c395dcea9f42aeda38bd4',): 2.8722813232690143,
 ('var-aggregate-d06292c54d071b6e5eb4558280d5246a',): 8.25}
```

**Comment by jcrist** [comment]
That is weird. I'll try to spin up a VM and take a look later today.

**Comment by mrocklin** [comment]
I got this running just by installing 32-bit miniconda on a 64-bit ubuntu system

**Comment by mrocklin** [comment]
Also, tests pass with the current fix.   I'm content with letting this lie if you're otherwise occupied.  I just wanted to raise it because it's, well, odd.

**Comment by jcrist** [comment]
Fine by me. I'll still try to take a look because I'm curious. There may be a followup PR with a more informed fix.

**Pull request merged**